### ISSUE TITLE
reportUninitializedInstanceVariable shouldn't be issued on ABCs

### ISSUE THREAD

**Issue by adampauls** [microsoft/pyright#5009]
Note: if you are reporting a wrong signature of a function or a class in the standard library, then the typeshed tracker is better suited for this report: https://github.com/python/typeshed/issues.

**Describe the bug**
`reportUninitializedInstanceVariable` issues an error even on ABCs (which aren't supposed to be instantiated). 

**To Reproduce**
```python
class Abstract(ABC):
    x: int  # error: Instance variable "x" is not initialized in the class body or __init__ method (reportUninitializedInstanceVariable)


@dataclass
class Concrete(Abstract):
    x: int = 5
```

Note that the error is already not issued on a `Protocol`:

```python
class Proto(Protocol):
    x: int  # pyright is ok with this
```

**Expected behavior**
Since ABCs aren't supposed to be instantiated directly (though of course it's python, so this one can be because it has no `@abstractmethod`s), I don't think this should be flagged as an error. In my dream world, `pyright` would issue an error on `Concrete` if it failed to initialize `x`, but that's probably asking a bit much.  

**VS Code extension or command-line**
Command-line with version 1.1.304. 

**Additional context**
Given the discussion in https://github.com/microsoft/pyright/issues/4701, I understand that this particular error is not the highest priority. Still, we have found it quite useful and hope it gets the love it deserves!

**Comment by erictraut** [comment]
Protocols are already exempt from this rule, so it makes sense to exempt subclasses of `ABC` (or any class that uses `ABCMeta` as its metaclass) as well.

This will be included in the next release.

**Comment by erictraut** [comment]
Correct, this change exempts all subclasses of ABC. Any such class can be abstract itself and further subclassed, so it's not an error to leave a variable uninitialized or a method unimplemented unless/until you attempt to instantiate it.

**Comment by erictraut** [comment]
I suppose I could "unexempt" such a class if it's marked as `@final` because it cannot be further subclassed.

**Comment by adampauls** [comment]
But you can subclass nonABCs too, so should it not also be fine to leave a variable uninitialized in that case either? I would argue that if you *can* instantiate a class, it should be considered a bug to declare but not implement a variable. It's a little surprising that if I take an existing class and add a super class to it (that indirectly inherits ABC), that would switch off the error for an uninitialized variable. 

```python
class WillErr:
  x: int

class WontError(SomethingHere):
  x: int
```

As always, thanks for the fantastic tool!

**Comment by erictraut** [comment]
All descendents of `ABC` in a class hierarchy are effectively abstract except for the final child, which is required to implement all of the remaining abstract methods and initialize all of the remaining unassigned variables. If you argue that the direct child of the `ABC` class should be exempt from this rule, then it follows that all descendents except for the final child should be exempt as well. The only way for a type checker to know that it's a final child is if you mark it as `@final`. I've implemented this the `@final` check, so you can use that if you'd like.

The "correct" behavior for the `reportUninitializedInstanceVariable` check is hard to pin down because Python really isn't designed for this coding pattern. That's why `reportUninitializedInstanceVariable` isn't enabled by default even in strict mode. I suspect that very few pyright users use it. I don't use it in any of my Python code. I'm glad to hear that you're getting some value out of it.

**Comment by adampauls** [comment]
> If you argue that the direct child of the ABC class should be exempt from this rule, then it follows that all descendents except for the final child should be exempt as well.

I see the disconnect. I am hoping for a world where `ABC` acts like `abstract` in Java/Scala: at every level of a hierarchy, you must declare yourself abstract or implement all methods. I know this is not how Python works, but I've been hoping that `pyright` can make it ever more so :) 

I've been using `ruff`, `pyright`, and PyCharm, all of which give different kinds of errors. I only just now realized that it is PyCharm, not `pyright`, that gives me the (IMHO helpful) error in this case:

```python
class Top(ABC):
    @abstractmethod
    def foo(self) -> int:
        ...

class Middle(Top):   # Class Middle must implement all abstract methods 
    ...


class Middle2(Top, ABC):   # no error
    ...
```

I would love it if `pyright` took the (opt-in) stance that you must declare yourself abstract or implement all methods and instantiate all variables, but its your tool and not mine. Thanks again for the quick feedback! The `@final` requirement is a reasonable one.

**Comment by erictraut** [comment]
> I would love it if pyright took the (opt-in) stance that you must declare yourself abstract or implement all methods and instantiate all variables

That's not consistent with the way Python works. It's also not consistent with other tools like mypy. So I don't think it would be appropriate for pyright to enforce this behavior.

**Comment by adampauls** [comment]
Right, I agree it should definitely want it to be opt-in because it's much tighter than what Python expects -- and of course that you've been helpful enough for one day!

Interestingly, pyright already catches the bug here 
```python
class Top(ABC):
    @abstractmethod
    def foo(self) -> int:
        ...


class Middle(Top):
    ...


Middle()  #  error: Cannot instantiate abstract class "Middle"
```

So I guess arguably, the consistent thing would be to detect uninitialized variables at constructor call site instead?

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.305, which I just published. It will also be included in a future release of pylance.
